### DISCUSSION TITLE
KIM1 connection to ESP32

### DISCUSSION THREAD

**Discussion by polperez22** [micropython/micropython#13435] [ESP32]
Hi there,
I'm trying to send data with my KIM1 shield v3 with my ESP32 and I'm having some trouble. I'm using UART module to communicate with the KIM1 but I don't get any response so I'm guessing it's not well communicated. 
Does someone know if I should be using a different module for my KIM1 or how can I get to send my data?
Thanks

**Comment by scruss** [comment]
Please, document/link what you mean as a "KIM1".

To me, it's a very old 6502 cpu trainer made by MOS (then later by Commodore) that uses high voltage (non-TTL) RS-232 serial communication

**Reply by polperez22** [reply]
My bad sorry. Here you have the link to the oficial product: https://connect.kineis.com/kim1/kim1

**Reply by davefes** [reply]
Thought I had a good look around their website ... where is a datasheet for this device?  

I see it is a 5V device have you got the voltage levels setup correctly to interface to this device with an ESP32?

**Reply by davefes** [reply]
Is this the device:
https://telemetry.groupcls.com/argos-solutions/argos-products/modems/kim1-interface-board/

**Reply by polperez22** [reply]
Yes it's that device. On the link I sent you you can have the datasheet and the integration manual.

**Reply by polperez22** [reply]
Here's the datasheet: https://static.helpjuice.com/helpjuice_production/uploads/upload/image/10644/direct/1658412966976-KINEIS-SP-20-0147%20KIM1%20Datasheet%20v2.0.pdf. You can use this as the differences between both versions are so light.

**Comment by davefes** [comment]
I expect 100-200 pages for a datasheet.  I found a one page datasheet.

**Reply by davefes** [reply]
That is better.

**Reply by davefes** [reply]
3.2.1 UART looks like 3V3 is OK, as well as 5V.

Where are all the AT commands listed?  How are you 'talking' to the device during development?

I will try to find the integration manual.

**Reply by polperez22** [reply]
That's where I'm blocked. My problem is that I can't find a way to "talk" to the device. Working with arduino ide is simpler bc this device has its own library. But with micropython there's not. So I tried with the machine library that includes UART but I haven't gotten to any results.

**Reply by polperez22** [reply]
Here you have the integration manual for KIM1: https://static.helpjuice.com/helpjuice_production/uploads/upload/image/10644/direct/1658413291767-KINEIS-NT-19-0018%20Kim1%20Integration%20Manual%20v2.3.pdf

**Reply by davefes** [reply]
Got it ... so when you issue the AT command:
`AT+ID=?`
what you do receive?

**Reply by davefes** [reply]
Show us how you are using the UART.

**Reply by davefes** [reply]
> Working with arduino ide is simpler bc this device has its own library.

Reading through it.

**Reply by davefes** [reply]
Do you have it connected to the ESP32 as per Fig 1 in the Integration Manual?

**Reply by davefes** [reply]
Have you done a loop-back test to verify that the ESP32 UART is working?

**Reply by davefes** [reply]
Connect TXD and RXD together on the ESP32.

```py
     #  do a loop-back test first

        msg = 'Test'
        ser.write(msg)
        print('message written')

        while True:
            new_msg = ser.read()
            if not  new_msg:
                break

        print(new_msg + '\n')
```

Back in a few hours.

**Reply by davefes** [reply]
Looking at the Arduino library line 188 in KIM.cpp should be fairly easy to port to Micropython.  Not that I understand the for loops.  It looks like you have to send the command multiple times??  `sendATCommand` will be a bigger porting job.

Good luck!
